File: common/src/main/java/me/pandamods/pandalib/config/api/holders/ConfigHolder.java
package me.pandamods.pandalib.config.api.holders;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import dev.architectury.platform.Platform;
import me.pandamods.pandalib.config.api.Config;
import me.pandamods.pandalib.config.api.ConfigData;
import me.pandamods.pandalib.utils.ClassUtils;
import net.minecraft.resources.ResourceLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;

public class ConfigHolder<T extends ConfigData> {
	public final Logger logger;
	private final Gson gson;

	private final Class<T> configClass;
	private final Config definition;
	private final ResourceLocation resourceLocation;
	private final ResourceLocation configPacketId;
	private final boolean synchronize;
	private T config;

	public ConfigHolder(Class<T> configClass, Config config) {
		this.configClass = configClass;
		this.definition = config;
		this.logger = LoggerFactory.getLogger(config.modId() + " | Config");
		this.gson = getNewDefault().buildGson(new GsonBuilder()).setPrettyPrinting().create();

		this.resourceLocation = new ResourceLocation(config.modId(), config.name());
		this.synchronize = config.synchronize();
		this.configPacketId = new ResourceLocation(config.modId(), "config_packet");

		this.load();
	}

	public Gson getGson() {
		return gson;
	}

	public Class<T> getConfigClass() {
		return configClass;
	}

	public Config getDefinition() {
		return definition;
	}

	public Path getConfigPath() {
		Path path = Platform.getConfigFolder();
		if (!definition.parentDirectory().isBlank()) path = path.resolve(definition.parentDirectory());
		return path.resolve(definition.name() + ".json");
	}

	public void save() {
		Path configPath = getConfigPath();
		try {
			Files.createDirectories(configPath.getParent());
			BufferedWriter writer = Files.newBufferedWriter(configPath);
			this.getGson().toJson(config, writer);
			writer.close();
			logger.info("successfully saved config '{}'", name());
		} catch (IOException e) {
			logger.info("Failed to save config '{}'", name());
			throw new RuntimeException(e);
		}
	}

	public boolean load() {
		Path configPath = getConfigPath();
		if (Files.exists(configPath)) {
			try (BufferedReader reader = Files.newBufferedReader(configPath)) {
				this.config = this.getGson().fromJson(reader, configClass);
			} catch (IOException e) {
				logger.error("Failed to load config '{}', using default", name(), e);
				resetToDefault();
				return false;
			}
		} else {
			resetToDefault();
		}
		logger.info("successfully loaded config '{}'", name());
		return true;
	}

	public void resetToDefault() {
		this.config = getNewDefault();
	}

	public T getNewDefault() {
		return ClassUtils.constructUnsafely(configClass);
	}

	public ResourceLocation resourceLocation() {
		return resourceLocation;
	}

	public String name() {
		return getDefinition().name();
	}

	public T get() {
		return config;
	}
}

File: src/main/kotlin/dev/pandasystems/pandalib/config/holders/ConfigHolder.kt
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package dev.pandasystems.pandalib.config.holders

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import dev.pandasystems.pandalib.config.Config
import dev.pandasystems.pandalib.config.ConfigData
import dev.pandasystems.pandalib.platform.Services
import dev.pandasystems.pandalib.utils.ClassUtils
import net.minecraft.resources.ResourceLocation
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.exists

open class ConfigHolder<T : ConfigData>(val configClass: Class<T>, val definition: Config) {
	val logger: Logger = LoggerFactory.getLogger(definition.modId + " | Config")
	val gson: Gson = newDefault.buildGson(GsonBuilder()).setPrettyPrinting().create()

	private val resourceLocation: ResourceLocation = ResourceLocation.fromNamespaceAndPath(definition.modId, definition.name)
	private val synchronize: Boolean = definition.synchronize
	private lateinit var config: T

	fun shouldSynchronize(): Boolean {
		return synchronize
	}
	
	val configPath: Path = Services.GAME.configDir.let { 
		if (definition.directory.isEmpty()) it else it.resolve(definition.directory)
	}.resolve(definition.name + ".json")
	
	init {
		if (load()) {
			save()
		}
	}

	open fun save() {
		val jsonObject = gson.toJsonTree(config).getAsJsonObject()
		this.config.onSave(this, jsonObject)
		
		val configPath = this.configPath
		try {
			Files.createDirectories(configPath.parent)
			val writer = Files.newBufferedWriter(configPath)
			this.gson.toJson(jsonObject, writer)
			writer.close()
			this.logger.info("Successfully saved config '{}'", definition.name)
		} catch (e: IOException) {
			this.logger.info("Failed to save config '{}'", definition.name)
			throw RuntimeException(e)
		}
	}

	open fun load(): Boolean {
		val configPath = this.configPath
		if (configPath.exists()) {
			try {
				Files.newBufferedReader(configPath).use { reader ->
					val jsonObject = this.gson.fromJson<JsonObject>(reader, JsonObject::class.java)
					this.config = this.gson.fromJson(jsonObject, configClass)
					this.config.onLoad(this, jsonObject)
				}
			} catch (e: IOException) {
				this.logger.error("Failed to load config '{}', using default", definition.name, e)
				resetToDefault()
				return false
			}
		} else {
			resetToDefault()
			save()
		}
		this.logger.info("Successfully loaded config '${definition.name}'")
		return true
	}

	fun resetToDefault() {
		this.config = this.newDefault
	}

	val newDefault: T
		/**
		 * @return Newly created class
		 */
		get() = ClassUtils.constructUnsafely(configClass)

	fun resourceLocation(): ResourceLocation {
		return resourceLocation
	}

	val langName: String
		get() = String.format("config.%s.%s", resourceLocation.namespace, resourceLocation.path)

	fun modID(): String {
		return this.definition.modId
	}

	/**
	 * @return Local config settings
	 */
	open fun get(): T? {
		return config
	}
}
